### PR TITLE
FOGL-6698 Add database tuning section

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Signed-off-by: Mark Riddoch <mark@dianomic.com>

### CODE REVIEW

File: docs/tuning_fledge.rst [hunk 1]
```diff
@@ -99,3 +104,149 @@ The Fledge core contains a scheduler that is used for running periodic tasks, th
...
+
+  - **Service Port**: Normally the storage service will dynamically create a service port that will be used by the storage service. Setting this to a value other than 0 will cause a fixed port to be used. This can be useful when developing a new storage plugin or to allow access to a non-fledge application to the storage layer. This should only be changed with extreme caution.
+
+  - **Management Port**: Normally the storage service will dynamically create a management port that will be used by the storage service. Setting this to a value other than 0 will cause a fixed port to be used. This can be useful when developing a new storage plugin.
+
+Changing will be saved once the *save* button is pressed. Fledge uses a mechanism whereby this data is not only saved in the configuration database, but also cached to a file called *storage.json* in the *etc* directory of the data directory. This is required such that Fledge can find the configuration database during the boot process. If the configuration becomes corrupt for some reason simply removing this file and restarting Fledge will cause the default configuration to be restored.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:163]: > data directory.

 if explicit data is set then FLEDGE_DATA directory else FLEDGE_ROOT/data directory.

Also we have an outstanding issue when we directly edit the storage.json file, GUI does not come up with right value of storage engine as configured in storage.json. See FOGL-6023

File: docs/tuning_fledge.rst [hunk 2]
```diff
@@ -99,3 +104,149 @@ The Fledge core contains a scheduler that is used for running periodic tasks, th
...
+
+  - **No. Readings per database**: The *sqlite* plugin support multiple readings databases, with the name of the asset used to determine which database to store the readings in. This improves the level of parallelism by reducing the lock contention when data is being written. Setting this value to 1 will cause only a single asset name to be stored within a single readings database, resulting in no contention between assets. However there is a limit on the number of databases, therefore setting this to 1 will limit the number of different assets that can be ingested into the instance.
+
+  - **No. databases to allocate in advance**: This controls how many reading databases Fledge should initially created. Creating databases is a slow process and thus is best achieved before data starts to flow through Fledge. Setting this too high will cause Fledge to allocate a large number of databases than required and waste open database connections. Ideally set this to the number of different assets you expect to ingest divided by the number of readings per database configuration above. This should give you sufficient databases to store the data you require.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:188]: Can we say here - default is with 3 database when Fledge is UP?

File: docs/tuning_fledge.rst [hunk 3]
```diff
@@ -99,3 +104,149 @@ The Fledge core contains a scheduler that is used for running periodic tasks, th
...
+
+   The *sqlite* configuration is still present and selectable since this instance has run that storage plugin in the past and the configuration is preserved when switching between *sqlite* and *sqlitelb* plugins.
+
+  - **Pool Size**: The storage service uses a connection pool to communicate with the underlying database, it is this pool size that determines how many parallel operations can be invoked on the database.
+
+    .. note::
+
+        Although the pool size denotes the number of parallel operations that can take place, database locking considerations may reduce the number of actual operations in progress at any point in time.
+
+    This pool size is only the initial size, the storage service will grow the pool if required, however setting a realistic initial pool size will improve the ramp up performance of Fledge.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:215]: Left indent these lines.
Otherwise it will display under L#205 note

**Review by ashish-jabble — COMMENTED** [review]
Also we can add cross reference target `Pool Size` and call this target in sqlite, sqlitelb, Postgres, sqlitememory configuration.
Following a DRY concept.

```
.. _Pool Size:

Pool Size
---------


The storage service uses a connection pool to communicate with the underlying database, it is this pool size that determines how many parallel operations can be invoked on the database.

     .. note::

         Although the pool size denotes the number of parallel operations that can take place, database locking considerations may reduce the number of actual operations in progress at any point in time.

     This pool size is only the initial size, the storage service will grow the pool if required, however setting a realistic initial pool size will improve the ramp up performance of Fledge.
```

And call above target with 
```
See :ref:`Pool Size`: for details.
```

File: docs/tuning_fledge.rst (hunk 2 above)
Comment (MarkRiddoch) [reply docs/tuning_fledge.rst:188]: I would rather not include the defaults here as they may get changed and then the documentation is out of date. I think it is obvious when you go an look that you see this is set to 3.

File: docs/tuning_fledge.rst [hunk 4]
```diff
@@ -99,3 +104,149 @@ The Fledge core contains a scheduler that is used for running periodic tasks, th
...
+
+Choosing A Storage Plugin
+-------------------------
+
+Fledge comes with a number of storage plugins that may be used, each one has it benefits and limitations, below is an overview of each of the plugins that are currently included with Fledge.
+
+sqlite
+    The default storage plugin that is used. It is implemented using the *SQLite* database and is capable of storing both configuration and reading data. It is optimized to allow parallelism when multiple assets are being ingested into the Fledge instance. It does however have limitations on the number of different assets that can be ingested within an instance. This is a good general purpose storage plugin and can manage reasonably high rates of data reading.
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply docs/tuning_fledge.rst:127]: It's hard to give an exact number, for lots of reasons, but yes, an indication would be good.

[the remark this answers is not in the document]
Comment (ashish-jabble) [reply docs/tuning_fledge.rst]: Thanks for the consideration.

File: docs/tuning_fledge.rst (hunk 2 above)
Comment (ashish-jabble) [reply docs/tuning_fledge.rst]: sure

**Review by ashish-jabble — APPROVED** [review]

**Pull request merged**